File: test/unit/cpp/micro_demag.cpp
#include "../../../src/mesh.cpp"
#include "../../../src/nonequispaced_mesh.cpp"//TODO needed for vtk_IO, include headers instead?
#include "../../../src/llg_terms/micro_demag.cpp"
#include "../../../src/integrators/new_llg.cpp"
#include "../../../src/integrators/adaptive_runge_kutta.cpp"
#include "../../../src/integrators/controller.cpp"
#include "../../../src/func.cpp"
#include "../../../src/state.cpp"
#include "../../../src/misc.cpp"
#include "../../../src/vtk_IO.cpp"
#include <gtest/gtest.h>
using namespace magnumafcpp;

TEST(MicroDemag, EnergyOfHomogeneousCube) {
    const double x=1.e-9, y=1.e-9, z=1.e-9;
    const int nx = 10, ny=10 , nz=10;

    Mesh mesh(nx, ny, nz, x/nx, y/ny, z/nz);
    double Ms = 1e5;

    af::array m = af::constant(0.0, mesh.n0, mesh.n1, mesh.n2, 3, f64);
    m(af::span, af::span, af::span, 0) = 1;
    State state(mesh, Ms, m);

    LLGIntegrator llg(1., {LlgTerm (new DemagField(mesh))});

    std::cout.precision(24);
    double llgE = llg.E(state);
    double analytic = 1./6. * x * y * z * pow(state.Ms, 2) * constants::mu0;

    EXPECT_NEAR(llgE, analytic, (llgE+analytic)/2. * 5e-8);// opencl precision
    //EXPECT_NEAR(llgE, analytic, (llgE+analytic)/2. * 1e-12);// cpu precision
}

int main(int argc, char **argv) {
    testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: python/test/unit/micro_demag.py
import unittest
import arrayfire as af
import numpy as np
import magnumaf
import math

class MicroDemagTest(unittest.TestCase):
    nx=10
    dx=1.e-10
    def test_micro_demag_homogen_cube(self):
        mesh=magnumaf.Mesh(self.nx, self.nx, self.nx, self.dx, self.dx, self.dx)
        m=af.constant(0.0, self.nx, self.nx, self.nx, 3, dtype=af.Dtype.f64)
        m[:, :, :, 0]=1.
        state=magnumaf.State(mesh, Ms = 1e5, m = m)
        micro_demag=magnumaf.DemagField(mesh)
        llg=magnumaf.LLGIntegrator(alpha = 0, terms = [micro_demag])
        self.assertAlmostEqual(llg.E(state), 1./6. * (self.nx*self.dx)**3 * state.Ms**2 * magnumaf.Constants.mu0)

if __name__ == '__main__':
  unittest.main()
